sdk/src/account.rs

```diff
@@ -764,6 +767,83 @@ pub const PROGRAM_OWNERS: &[Pubkey] = &[
     loader_v4::id(),
 ];
 
+const LOADER_V4_STATUS_BYTE_OFFSET: usize = 40;
+
+/// Create executable account meta data based on account's `owner`.
+///
+/// This function is only used for testing and an optimization during
+/// transaction loading.
+///
+/// When the program account is already present in the program cache, we don't
+/// need to load the full account data during transaction loading. Instead, all
+/// we need is a minimal executable account meta data, which is what this
+/// function returns.
+pub fn create_executable_meta(owner: &Pubkey) -> &[u8] {
+    // For upgradable program account, only `UpgradeableLoaderState::Program`
+    // variant (i.e. discriminant = 2) should *executable*, which means the
+    // discriminant for the enum at byte offset 0 in account data is 2.
+    const EXECUTABLE_META_FOR_BPF_LOADER_UPGRADABLE: [u8; 1] = [2];
+
+    // For loader v4 program, when LoaderV4Status (byte_offset = 40 in account
+    // data) is set, the program is executable.
+    const fn get_executable_meta_for_loader_v4() -> [u8; 41] {
+        let mut v = [0; LOADER_V4_STATUS_BYTE_OFFSET + 1];
+        v[LOADER_V4_STATUS_BYTE_OFFSET] = 1;
+        v
+    }
+    const EXECUTABLE_META_FOR_LOADER_V4: [u8; LOADER_V4_STATUS_BYTE_OFFSET + 1] =
+        get_executable_meta_for_loader_v4();
+
+    // For other owners, simple returns a 1 byte array would make it executable.
+    const DEFAULT_EXECUTABLE_META: [u8; 1] = [1];
+
+    if bpf_loader_upgradeable::check_id(owner) {
+        &EXECUTABLE_META_FOR_BPF_LOADER_UPGRADABLE
+    } else if loader_v4::check_id(owner) {
+        &EXECUTABLE_META_FOR_LOADER_V4
+    } else {
+        &DEFAULT_EXECUTABLE_META
+    }
+}
+
+/// Return true if the account program is executable.
+pub fn is_executable(account: &impl ReadableAccount) -> bool {
+    // empty account is not executable.
+    if account.data().is_empty() {
+        return false;
+    }
+
+    // bpf_loader still relies on `executable` on the program account. When the
+    // program account is finalized, the loader will mark `executable` flag on
+    // the account. Once we disable the deployment of these two loaders, we can
+    // remove the following dependency on `executable`.
+    if bpf_loader::check_id(account.owner()) || bpf_loader_deprecated::check_id(account.owner()) {
+        return account.executable();
```

Comment:
`return true;`
That will be safe once deployments on loader-v2 are disabled. And we don't switch to this emulation here before that feature is activated.